packages/datadog-instrumentations/src/router.js

```diff
@@ -84,10 +86,15 @@ function createWrapRouterMethod (name) {
   }
 
   function wrapNext (req, next) {
+    if (next[DD_WRAPPED_SYMBOL]) {
+      return next
+    }
     return shimmer.wrapFunction(next, next => function (error) {
       if (error && error !== 'route' && error !== 'router') {
         errorChannel.publish({ req, error })
       }
+      // ensure we don't wrap the same next function twice
+      next[DD_WRAPPED_SYMBOL] = true
```

Comment:
The wrapping happens as soon as `wrapFunction()` is called. So in case the method is executed later, we could have still instrumented it multiple times, if we don't immediately set the symbol to prevent a second instrumentation.

I would also use a WeakSet to prevent manipulating the next code.

```js
const instrumentedNextInstances = new WeakSet()
```

```suggestion
    // Ensure we don't wrap the same next function twice
    if (instrumentedNextInstances.has(next)) {
      return next
    }
    instrumentedNextInstances.add(next)

    return shimmer.wrapFunction(next, next => function (error) {
      if (error && error !== 'route' && error !== 'router') {
        errorChannel.publish({ req, error })
      }
```